### PR TITLE
raftstore: achieve the accuracy of latency inspect record when write sync

### PR DESCRIPTION
**Pull request by JayiceZ**
refer https://github.com/tikv/tikv/issues/12254
Signed-off-by: Jayice <1185430411@qq.com>

### What is changed and how it works?

Issue Number: ref https://github.com/tikv/tikv/issues/12254

What's Changed:

```
the original logic will double count dur when write synchronously
this PR can fix it.
```

### Check List 

Tests 

- Integration test

### Release note 

```release-note
none
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/store.rs [hunk 1]
```diff
@@ -841,20 +841,56 @@ impl<EK: KvEngine, ER: RaftEngine, T: Transport> PollHandler<PeerFsm<EK, ER>, St
...
+        if let Some(write_worker) = &mut self.poll_ctx.sync_write_worker {
+            if self.poll_ctx.has_ready {
+                let write_begin = TiInstant::now();
                 write_worker.write_to_db(false);
+                let write_finish = TiInstant::now();
+
+                if !latency_inspect.is_empty() {
+                    for mut inspector in latency_inspect {
+                        inspector.record_store_write(
+                            write_finish.saturating_duration_since(write_begin),
```
Comment (5kbpers) [inline components/raftstore/src/store/fsm/store.rs:866]: How about `write_begin.saturating_elapsed()`?

File: components/raftstore/src/store/fsm/store.rs [hunk 2]
```diff
@@ -841,20 +841,56 @@ impl<EK: KvEngine, ER: RaftEngine, T: Transport> PollHandler<PeerFsm<EK, ER>, St
...
                 }
+            } else {
+                for inspector in latency_inspect {
+                    inspector.finish();
+                }
             }
-
-            let dur = self.timer.saturating_elapsed();
+        } else {
+            let now = TiInstant::now();
```
Comment (5kbpers) [inline components/raftstore/src/store/fsm/store.rs:881]: We can reuse `write_begin` here.

**Review by 5kbpers — COMMENTED** [review]
rest LGTM

File: components/raftstore/src/store/fsm/store.rs (hunk 1 above)
Comment (JayiceZ) [reply components/raftstore/src/store/fsm/store.rs]: Good suggestion! Fixed

File: components/raftstore/src/store/fsm/store.rs (hunk 2 above)
Comment (JayiceZ) [reply components/raftstore/src/store/fsm/store.rs]: Fixed~

**Comment by JayiceZ** [comment]
Oh, Can anyone else PTAL?

**Comment by JayiceZ** [comment]
@Connor1996 PTAL if you have time~

File: components/raftstore/src/store/fsm/store.rs [hunk 3]
```diff
@@ -841,20 +841,52 @@ impl<EK: KvEngine, ER: RaftEngine, T: Transport> PollHandler<PeerFsm<EK, ER>, St
...
+        if self.poll_ctx.has_ready {
             // Only enable the fail point when the store id is equal to 3, which is
             // the id of slow store in tests.
             fail_point!("on_raft_ready", self.poll_ctx.store_id() == 3, |_| {});
+        }
+        let mut latency_inspect = std::mem::take(&mut self.poll_ctx.pending_latency_inspect);
+        let mut dur = self.timer.saturating_elapsed();
 
-            if let Some(write_worker) = &mut self.poll_ctx.sync_write_worker {
+        if !latency_inspect.is_empty() {
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/store.rs:852]: no need

File: components/raftstore/src/store/fsm/store.rs [hunk 4]
```diff
@@ -841,20 +841,52 @@ impl<EK: KvEngine, ER: RaftEngine, T: Transport> PollHandler<PeerFsm<EK, ER>, St
...
+            }
+        }
+        let write_begin = TiInstant::now();
+        if let Some(write_worker) = &mut self.poll_ctx.sync_write_worker {
+            if self.poll_ctx.has_ready {
                 write_worker.write_to_db(false);
 
+                if !latency_inspect.is_empty() {
+                    for mut inspector in latency_inspect {
+                        inspector.record_store_write(write_begin.saturating_elapsed());
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/store.rs:864]: Not sure if the original logic is on purpose, as the waterfall metrics may be accumulative? /cc @gengliqi @innerr

**Review by gengliqi — APPROVED** [review]
LGTM

File: components/raftstore/src/store/fsm/store.rs (hunk 3 above)
Comment (JayiceZ) [reply components/raftstore/src/store/fsm/store.rs]: Removed. Thanks

**Review by 5kbpers — APPROVED** [review]

**Pull request merged**